### PR TITLE
Add custom validator to blueprint variables

### PR DESCRIPTION
**Pull request by phobologic**
- Now all blueprint variables must provide a `type`, otherwise an
  exception is thrown.

Before it wasn't required, and the type was ignored if it wasn't
provided. That didn't seem right to me, but I realize I may just not
have understood the implications.

Other changes:

- variable resolution now done in a function not a class method
- moved variable type validation to a function
- instead of passing the whole blueprint just to use it's name, I just
  pass the blueprints name into the exceptions that use them (that was
  all the blueprint was used for it appeared)

I'll add docs in the next checkin.

### CODE REVIEW

**Comment by phobologic** [comment]
@mhahn @jeffpierce

**Comment by phobologic** [comment]
Fixes #217

File: stacker/blueprints/base.py [hunk 1]
```diff
@@ -130,6 +131,90 @@ def build_parameter(name, properties):
...
+        # Variable value not provided, try using the default, if it exists
+        # in the definition
+        try:
+            value = var_def["default"]
+        except KeyError:
+            raise MissingVariable(blueprint_name, var_name)
+
+    # If no validator, return the value as is, otherwise apply validator
+    validator = var_def.get("validator", lambda v: v)
+    value = validator(value)
```
Comment (mhahn) [inline stacker/blueprints/base.py]: I think we should have a preferred exception that validators raise that gets caught here. Then we can include more details about which variable was invalid etc.

Right now if you raise an exception within the validator, you can't tell which variable was being validated.

Comment (mhahn) [inline stacker/blueprints/base.py]: or maybe just any exception that is raised by the validator we wrap with some other info?

Comment (phobologic) [inline stacker/blueprints/base.py]: Good call - I'll think about it and come up with something that makes it easy to tell which variable failed it's validator, and why.

File: stacker/blueprints/base.py [hunk 2]
```diff
@@ -130,6 +132,93 @@ def build_parameter(name, properties):
...
+            value = var_def["default"]
+        except KeyError:
+            raise MissingVariable(blueprint_name, var_name)
+
+    # If no validator, return the value as is, otherwise apply validator
+    validator = var_def.get("validator", lambda v: v)
+    try:
+        value = validator(value)
+    except Exception as exc:
+        raise ValidatorError(var_name, validator.__name__, value, exc)
```
Comment (mhahn) [inline stacker/blueprints/base.py]: add this to docstring above

**Comment by mhahn** [comment]
👍

**Pull request merged**